Feature/coprocessing

m3d opened the pull request:
This is the first attempt for parallel processing (on separate core using multiprocessing). The solution is so far very ugly and the commits will be surely squashed.

zwn commented:
@m3d Please give me a full command line how to execute this (including logname). Thanks.

zwn commented:
I have tried zip-logs, but reading images fails there. I have tried some non-zip older log I have here (used to get camera calibration) but that fails with assert.

zwn commented:
What I get from just reading the code.

I'd use process pools described in [python multiprocessing docs](https://docs.python.org/2/library/multiprocessing.html#using-a-pool-of-workers). I guess `apply_async` is the method you are looking for. In that case you don't need the queues at all. When an image is available you just call `pool.apply_async(process_fn, (image,))` and keep a list of results of submitted computations. You can check with `res.ready()` if the computation is done whenever you advance to the next step in the main program.

m3d commented:
Here is an example of test from log file:
```
python.exe navpat.py d:\md\osgar\logs\czu170620\osgar_170620_175322.zip
```
Note, that in that case processing is skipped and the results are only read from the old log. Testing workaround is to change https://github.com/robotika/osgar/pull/9/files#diff-702b6abf015bfdb4e41308ddd3b297a9R105 to `if False:` and unzip the log to separate directory. Then only the processing will run while other inputs will be loaded from the log file.

m3d commented:
The solution with `multiprocessing.Pool` and 'apply_async' looks nice, thanks :). I would try it (probably as extra commit).

m3d commented:
Thanks for hints - it looks much nicer now and so far it also seems to be working even as `self.pool`, i.e. without need for globals :+1: .

zwn commented on lib/processor.py:
```diff
@@ -0,0 +1,35 @@
+        self.processing = None
+
+    def get_result(self):
+        if self.processing is None or self.processing.ready():
```
if self.processing is None or __not__ self.processing.ready()?

m3d commented:
I would not merge this branch to feature/improve-navpat but later rebase it on master.

m3d commented:
I squashed the code and rebased it on master. At the moment I am aware of bug in the case of unused parallel processing - similar case occurred if you do not attach camera for example.
The code can be tested from log via
```
python.exe navpat.py d:\md\osgar\logs\czu170627\czuC\meta_170627_173554.log --view F
```
Note, that the zipped logs are not merged into master yet.

Also it would be useful to be able to recompute the _processing_ separately, without original `navpat.py`. For that it is necessary to know callback function (in this pull-request is simply computation of green pixels on left and right side), inputs and for verification outputs. In this particular example the output contains also original input - should this be mandatory?

zwn commented:
> At the moment I am aware of bug in the case of unused parallel processing - similar case occurred if you do not attach camera for example.

The way you described your intentions I would think that it is by design and not a bug. IE. the only way to replay the log is to do everything the same.

I am able to replay the log file https://osgar.robotika.cz/osgar_170627_173554.zip.

> Also it would be useful to be able to recompute the processing separately, without original navpat.py. 

To be able to do the same processing to verify the outputs, you would need much more than the name of the callback - you would need the sha1 of the master + complete diff of the feature branch + name of the callback. The first part is needed even for replay using navpat.py. Let's just not go there yet.

Feel free to merge.

zwn commented:
One more question - why doesn't it replay without "F"?

m3d commented:
In order to replay the log file https://osgar.robotika.cz/osgar_170627_173554.zip wihtout "F" you would need
corresponding source code which is with different steering offset and start position (feature/test170627).

The code should not crash for any inputs and it will if you use old files without Processor, i.e. files from "garden party".
```
Traceback (most recent call last):
  File "navpat.py", line 262, in <module>
    navigate_pattern(metalog, viewer)
  File "navpat.py", line 238, in navigate_pattern
    detach_all_sensors(robot)
  File "m:\git\osgar\helper.py", line 131, in detach_all_sensors
    robot.proc.requestStop()
AttributeError: 'JohnDeere' object has no attribute 'proc'
```

```
python.exe navpat.py d:\md\osgar\logs\czu170610\tmp\meta_170608_133319.log F
```

Maybe we should refactor `robot.data_sources` and `robot.threads` into dictionary with common API?

m3d commented:
Is also the last (after review) commit "approved"? thanks

zwn commented:
yes :+1:
